### ISSUE TITLE
RAMANTRANSFORMER and SANET collapse to constant (DUMMY-identical) predictions on several datasets

### ISSUE THREAD

**Issue by mario-koddenbrock** [ml-lab-htw/RamanBench#10]
Found while verifying a claim in #7. Checking `data/precomputed/classification_metrics.csv` directly: `RAMANTRANSFORMER` and `SANET` match `DUMMY`'s score exactly (f1_score and balanced_accuracy identical to 6 decimal places) across all 3 seeds, on datasets where other models do *not* collapse — meaning this looks like a genuine training failure for these two architectures specifically, not a shared unlearnable-target artifact.

**RAMANTRANSFORMER exact-DUMMY-match datasets** (all 3 seeds):
- `Alzheimer's SERS Serum`
- `Saliva Parkinson`
- `Mutant Wheat`
- `Weathered Microplastics`
- `Diabetes Skin (Ear Lobe / Inner Arm / Thumbnail / Vein)`

**SANET exact-DUMMY-match datasets** (all 3 seeds):
- `Saliva Alzheimer`
- `Weathered Microplastics`
- `Diabetes Skin (Ear Lobe / Inner Arm / Thumbnail / Vein)`

The `Diabetes Skin (*)` matches are shared by several other models too (GBM, LR) and are most likely a genuinely unlearnable target, consistent with the benchmark's own learnability-draw handling — not in scope here. The datasets unique to RAMANTRANSFORMER/SANET are the actual signal: balanced_accuracy sits exactly at chance (0.5 for binary, 0.333 for 3-class), consistent with the model outputting a constant prediction rather than learning anything.

## Reproduction

```python
import pandas as pd
df = pd.read_csv('data/precomputed/classification_metrics.csv')
dummy = df[df['model']=='DUMMY'].set_index(['dataset','seed'])[['f1_score','balanced_accuracy']]
for (model, dataset), g in df[df['model']!='DUMMY'].groupby(['model','dataset']):
    if len(g) != 3:
        continue
    if all(
        abs(row.f1_score - dummy.loc[(row.dataset, row.seed), 'f1_score']) < 1e-9
        and abs(row.balanced_accuracy - dummy.loc[(row.dataset, row.seed), 'balanced_accuracy']) < 1e-9
        for row in g.itertuples()
    ):
        print(model, dataset)
```

## Suspected cause (not yet confirmed)

Both are custom deep learning architectures (`raman_bench/models/custom/ramantransformer`, `.../sanet`); a training collapse (e.g. dead ReLUs, an unstable learning rate for very small batch/class counts, or a normalization step producing NaNs that AutoGluon silently falls back on) on specific dataset shapes seems more likely than a shared data issue, since PLS/RF/other simpler models trained on the exact same folds don't collapse on these datasets.

## Suggested next step

Re-run just these two models on the flagged datasets with verbose/debug logging to see whether training loss actually decreases at all, or diverges/NaNs immediately.

**Comment by mario-koddenbrock** [comment]
Dug into this locally (no cluster needed — reproducible with just `raman_data` + the model code directly). Refined picture, splitting into three separate causes rather than one:

## 1. `Diabetes Skin (*)` — confirmed, not RAMANTRANSFORMER/SANET-specific

All four variants (`diabetes_skin_ear_lobe/inner_arm/thumbnail/vein`) have **exactly 20 samples total**. That's why GBM/LR collapse there too, not just the two deep models — this is a genuinely unlearnable-at-this-sample-size case, consistent with the benchmark's own learnability handling. Not a model bug; removing this from the "suspect" list.

## 2. RAMANTRANSFORMER on `Alzheimer's SERS Serum` / `Saliva Parkinson` / `Mutant Wheat` / `Weathered Microplastics` — likely root cause found

`RamanTransformerModel`'s default architecture is `d_model=768, nhead=12, n_layers=12` — a full ViT-Base-scale transformer (**85M parameters**, confirmed by instantiating it directly). Timed a single forward+backward pass on a 128-sample batch, 1748 features, CPU: **12.4s/batch**. With `batch_size=128`, that's ~4-5 minutes per epoch on a few-thousand-sample dataset like `wheat_lines` (53,134 samples total, well-balanced across 4 classes — ruling out small-N or imbalance as the cause here).

At that per-batch cost, any realistic AutoGluon `time_limit` (e.g. 3600s) allows roughly **10 epochs**, right at the model's own `patience=10` early-stopping threshold. Transformers trained from scratch (no pretraining here) are well known to sit at the "predict the marginal class distribution" solution for a number of epochs before escaping it — if that escape needs more than ~10 epochs, `patience`-based early stopping (or the outer time limit) locks in the collapsed state as "best `val_loss`," since early in training that trivial solution often *is* the lowest-loss checkpoint seen so far.

This would explain all four flagn datasets: none are tiny (1151-53134 samples), so it's the model's own architecture/budget mismatch, not the data.

**Not yet done**: haven't run the model to convergence with a much longer budget to positively confirm it "escapes" collapse given enough time — that's the natural next check, but each epoch's ~5 minutes on CPU makes a definitive run slow to do casually; happy to let this run in the background overnight, or try on GPU where it should be far cheaper per epoch.

## 3. SANET on `Saliva Alzheimer` — still open

SANet is much cheaper (**2.86M params, 1.9s/batch**) and `alzheimer` isn't small (1151 samples) or severely imbalanced (24%/76% split) — the compute-budget explanation doesn't obviously fit here. Haven't identified a cause yet. `Weathered Microplastics` (97 samples, several classes with n=1) is more likely a small-N case like #1 above, for both RAMANTRANSFORMER and SANET.

## Suggested next step

For #2: either shrink `RamanTransformerModel`'s default hyperparameters to something that can actually complete enough epochs within the real time budget (loses literature-fidelity to Liu et al. 2023, which was presumably run on GPU with a much longer schedule), or explicitly document that it needs a longer time budget / GPU to be meaningfully evaluated on CPU-tier jobs. That's a real design tradeoff, not something I'll change without input.

**Comment by mario-koddenbrock** [comment]
Correction to my last comment — the speed/budget theory is wrong. (Thanks for the sanity check that the cluster runs use an A100, not CPU — that sent me back to actually test this properly instead of reasoning from a CPU timing.)

Ran `RamanTransformerModel.fit()` on `wheat_lines` (4000-row subsample, 3200 train) to real convergence: 60 epochs, early stopping **disabled** (`patience=1000`), on MPS (~90 GPU-minutes total, decoupled from the CPU-speed argument since this removes the "not enough epochs" variable entirely).

Result — it never learns:

```
epoch   0  val_loss=1.4589  val_pred_dist=[0, 0, 480, 0]
epoch   1  val_loss=1.4321  val_pred_dist=[480, 0, 0, 0]
epoch   2  val_loss=1.5075  val_pred_dist=[0, 0, 480, 0]
epoch   3  val_loss=1.4821  val_pred_dist=[0, 0, 0, 480]
epoch   4  val_loss=1.4068  val_pred_dist=[0, 0, 0, 480]
...                                          <- locks here, all 480 predictions are class 3
epoch  59  val_loss=1.3836  val_pred_dist=[0, 0, 0, 480]

test preds: all 800 samples predicted class 3
test accuracy: 0.29  (== class 3's base rate, 232/800)
```

Epochs 0-3 flip between different constant-class predictions; from epoch 4 it locks onto one class and **never moves again for the remaining 55 epochs** despite early stopping being off. `train_loss`/`val_loss` both plateau at ~1.38 — essentially `ln(4) = 1.386`, the cross-entropy of producing no discriminative signal at all across 4 classes. This isn't "needs more epochs," it's a genuine optimization failure: the model settles into outputting the same non-informative logits regardless of input, this early, and stays there.

Likely candidates now: learning rate (`lr=1e-3` with plain Adam, no special init, for an 85M-parameter from-scratch transformer — many from-scratch ViT recipes use far smaller LR + much longer warmup than the 10 epochs here), or a depth-related init issue (12 layers, no depth-scaled init). Overparameterization relative to data is still plausible too (85M params against a few thousand training rows either way), but the completeness/speed of the collapse (locked by epoch 4, zero movement after) points more at optimization dynamics than pure data scarcity.

**Caveat**: this used a 4000-row subsample of `wheat_lines`, not the full 53,134-row dataset production likely used — haven't re-tested at full scale. Given how quickly and totally it locks up here, I'd be surprised if more data alone fixes it, but flagging honestly rather than overclaiming.

**Comment by mario-koddenbrock** [comment]
Ran the two suspects from my last comment (lower LR, GPT-2-style depth-scaled init) head-to-head against the baseline and their combination, then went further and tested model size directly, on the same `wheat_lines` 4000-row subsample (20 epochs each, patience disabled).

**All four LR/init configs collapse identically**: default (lr=1e-3, default init), lr=1e-4, scaled init, and both combined all plateau at `val_loss≈1.38` (≈ln(4)) and lock onto a single constant class — same 0.285 test accuracy across the board. Neither theory from my last comment holds up.

**Model size doesn't matter either.** Tested three architectures shrunk from the 85M-param default: 810K, 158K, and 29K params (2,900x smaller). Every single one collapses to the exact same pattern — same loss plateau, same locked single-class output, same 0.285 test accuracy. Overparameterization isn't the cause.

**Ruled out a harness/data bug as well.** A trivial single `nn.Linear` layer, trained on the exact same data/split with the exact same code, reaches 55% test accuracy in 15 epochs with no collapse — so the raw features genuinely carry learnable signal here, and this isn't an artifact of my test setup.

So this isn't a training-recipe problem (LR, init, or capacity) — it's something structural in `_RamanTransformerNetwork`'s architecture or forward pass, independent of depth. Two remaining suspects I haven't tested yet:
1. `nn.TransformerEncoderLayer` here defaults to **Post-LN** (no `norm_first=True`). Post-LN is well documented as harder to train from scratch than Pre-LN — though this sits a bit awkwardly with the 2-layer model also failing identically, since Post-LN instability is usually depth-dependent and a 2-layer Post-LN transformer is normally quite trainable.
2. Something in the patch-tokenization / CLS-token readout mechanism itself, independent of the attention blocks.

Haven't tested either yet — flagging both for whoever picks this up next.
